### PR TITLE
cpu/nrf5x: allow multiple I2C and SPI buses on a shared periph 

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

This allows to use the same hardware peripheral for different I2C / SPI busses when the `nrf5x_periph_shared` module is enabled. 

Since I2C and SPI share the peripheral, this means that now SPI0 can be used event when TWIM0 is used for I2C.
Or multiple I2C busses that use TWIM0 can be defined.

### Testing procedure

*This is entirely untested* 

Define several I2C or SPI busses that use the same hardware peripheral, but with different pins.
With `nrf5x_periph_shared` enabled, they should all work and re-mux as needed.

### CODE REVIEW

**Comment by nica-f** [comment]
I just tried it on nRF52840 with several I2C devices connected to different pin sets.
I have TWIM0 assigned to the CST816 touchscreen driver, all other I2C devices are assigned to TWIM1 using different pin setups in the board's periph_conf:

```
static const i2c_conf_t i2c_config[] = {
    {
        // CST816S touschscreen
        .dev = NRF_TWIM0,
        .scl = TOUCH_SCL,
        .sda = TOUCH_SDA,
        .speed = I2C_SPEED_FAST
    },
    {   // accellerometer KX023
        .dev = NRF_TWIM1,
        .scl = ACCEL_SCL,
        .sda = ACCEL_SDA,
        .speed = I2C_SPEED_FAST
    },
    {   // pressure/temp sensor BMP280
        .dev = NRF_TWIM1,
        .scl = ATM_PRESSURE_SCL,
        .sda = ATM_PRESSURE_SDA,
        .speed = I2C_SPEED_FAST
    },
    {   // heart rate sensor
        .dev = NRF_TWIM1,
        .scl = HRM_SCL,
        .sda = HRM_SDA,
        .speed = I2C_SPEED_FAST
    },
    {   // magnetometer
        .dev = NRF_TWIM1,
        .scl = MAGN_SCL,
        .sda = MAGN_SDA,
        .speed = I2C_SPEED_FAST
    },
};

#define I2C_NUMOF           ARRAY_SIZE(i2c_config)
```

Now all access to buses > 0 block indefinitely.

There are more devices on SPI but these use the buses > 1:

```
static const spi_conf_t spi_config[] = {
    {
        // LPM013M126 LCD
        .dev  = NRF_SPIM2,
        .sclk = LCD_SCK,
        .mosi = LCD_MOSI,
        //.miso = GPIO_UNDEF,
        //.ppi = 0,
    },
    {
        // SPI flash
        .dev  = NRF_SPIM3,
        .sclk = SPI_FLASH_SCK,
        .mosi = SPI_FLASH_MOSI,
        .miso = SPI_FLASH_MISO,
        //.ppi = 0,
    }
};

#define SPI_NUMOF           ARRAY_SIZE(spi_config)
```

File: cpu/nrf52/include/periph_cpu.h [hunk 1]
```diff
@@ -264,6 +264,34 @@ void spi_twi_irq_register_spi(NRF_SPIM_Type *bus,
 void spi_twi_irq_register_i2c(NRF_TWIM_Type *bus,
                               spi_twi_irq_cb_t cb, void *arg);
 
+/**
+ * @brief   Acquire the shared I2C/SPI peripheral in I2C mode
+ *
+ * @param   bus bus to acquire exclusive access on
+ */
+void nrf5x_i2c_acquire(NRF_TWIM_Type *bus);
```
Comment (maribu) [inline cpu/nrf52/include/periph_cpu.h:272]:
```suggestion
void nrf5x_i2c_acquire(NRF_TWIM_Type *bus, spi_twi_irq_cb_t cb, void *arg);
```

The shared peripherals have a single IRQ vector, which then needs to be dispatched to the correct ISR function. For this, the ISR will read `_irq[num]` and `_irq_arg[num]` for SPI/I2C peripheral number `num`. Just updating those on acquire may be sufficient.

**Comment by maribu** [comment]
My testing setup is this:

```diff
diff --git a/boards/common/nrf52/include/cfg_i2c_default.h b/boards/common/nrf52/include/cfg_i2c_default.h
index 6b59e974b8..54ba0a0cfe 100644
--- a/boards/common/nrf52/include/cfg_i2c_default.h
+++ b/boards/common/nrf52/include/cfg_i2c_default.h
@@ -32,10 +32,16 @@ extern "C" {
  */
 static const i2c_conf_t i2c_config[] = {
     {
-        .dev = NRF_TWIM1,
+        .dev = NRF_TWIM0,
         .scl = 27,
         .sda = 26,
         .speed = I2C_SPEED_NORMAL
+    },
+    {
+        .dev = NRF_TWIM1,
+        .scl = 30,
+        .sda = 31,
+        .speed = I2C_SPEED_NORMAL
     }
 };
 #define I2C_NUMOF           ARRAY_SIZE(i2c_config)
diff --git a/boards/nrf52840dk/include/periph_conf.h b/boards/nrf52840dk/include/periph_conf.h
index b615b9030c..b80a350a68 100644
--- a/boards/nrf52840dk/include/periph_conf.h
+++ b/boards/nrf52840dk/include/periph_conf.h
@@ -33,17 +33,11 @@ extern "C" {
  */
 static const spi_conf_t spi_config[] = {
     {
-        .dev  = NRF_SPIM0,
+        .dev  = NRF_SPIM2,
         .sclk = GPIO_PIN(1, 15),
         .mosi = GPIO_PIN(1, 13),
         .miso = GPIO_PIN(1, 14),
     },
-    {
-        .dev  = NRF_SPIM2,
-        .sclk = GPIO_PIN(0, 19),
-        .mosi = GPIO_PIN(0, 20),
-        .miso = GPIO_PIN(0, 21),
-    }
 };
 #define SPI_NUMOF           ARRAY_SIZE(spi_config)
 /** @} */
diff --git a/examples/saul/Makefile b/examples/saul/Makefile
index 7f2ca1e297..7b478a2d61 100644
--- a/examples/saul/Makefile
+++ b/examples/saul/Makefile
@@ -2,7 +2,7 @@
 APPLICATION = saul_example
 
 # If no BOARD is found in the environment, use this default:
-BOARD ?= native
+BOARD := nrf52840dk
 
 # This has to be the absolute path to the RIOT base directory:
 RIOTBASE ?= $(CURDIR)/../..
@@ -14,6 +14,8 @@ USEMODULE += shell
 USEMODULE += shell_commands
 # additional modules for debugging:
 USEMODULE += ps
+USEMODULE += bmp280_spi
+USEMODULE += sht3x
 
 # Comment this out to disable code in RIOT that does safety checking
 # which is not needed in a production environment but helps in the
@@ -24,3 +26,8 @@ DEVELHELP ?= 1
 QUIET ?= 1
 
 include $(RIOTBASE)/Makefile.include
+
+CFLAGS += '-DBMX280_PARAM_SPI=SPI_DEV(0)'
+CFLAGS += '-DBMX280_PARAM_CS=GPIO_PIN(1,10)'
+CFLAGS += '-DSHT3X_PARAMS={.i2c_dev=I2C_DEV(0),.i2c_addr=SHT3X_I2C_ADDR_1,.mode=SHT3X_PERIODIC_2_MPS,.repeat=SHT3X_HIGH},{.i2c_dev=I2C_DEV(1),.i2c_addr=SHT3X_I2C_ADDR_1,.mode=SHT3X_PERIODIC_2_MPS,.repeat=SHT3X_HIGH}'
+CFLAGS += '-DSHT3X_SAUL_INFO={.name="sht3x 1"},{.name="sht3x 2"}'
```

This does work even with `master`, as the two I2C and the SPI sensor are on different shared periphals. Now changing them all to SPIM0 should result in conflicts, that this PR should solve.

**Comment by maribu** [comment]
With this, it works:

```diff
diff --git a/cpu/nrf52/include/periph_cpu.h b/cpu/nrf52/include/periph_cpu.h
index d888c0a55c..9d6530845e 100644
--- a/cpu/nrf52/include/periph_cpu.h
+++ b/cpu/nrf52/include/periph_cpu.h
@@ -268,8 +268,10 @@ void spi_twi_irq_register_i2c(NRF_TWIM_Type *bus,
  * @brief   Acquire the shared I2C/SPI peripheral in I2C mode
  *
  * @param   bus bus to acquire exclusive access on
+ * @param   cb  ISR handler to call on IRQ
+ * @param   arg ISR handler argument
  */
-void nrf5x_i2c_acquire(NRF_TWIM_Type *bus);
+void nrf5x_i2c_acquire(NRF_TWIM_Type *bus, spi_twi_irq_cb_t cb, void *arg);
 
 /**
  * @brief   Release the shared I2C/SPI peripheral in I2C mode
@@ -281,16 +283,18 @@ void nrf5x_i2c_release(NRF_TWIM_Type *bus);
 /**
  * @brief   Acquire the shared I2C/SPI peripheral in SPI mode
  *
- * @param   bus bus to acquire exclusive access on
+ * @param   bus bus to release exclusive access on
+ * @param   cb  ISR handler to call on IRQ
+ * @param   arg ISR handler argument
  */
-void nrf5x_spi_release(NRF_SPIM_Type *bus);
+void nrf5x_spi_acquire(NRF_SPIM_Type *bus, spi_twi_irq_cb_t cb, void *arg);
 
 /**
  * @brief   Acquire the shared I2C/SPI peripheral in SPI mode
  *
- * @param   bus bus to release exclusive access on
+ * @param   bus bus to acquire exclusive access on
  */
-void nrf5x_spi_acquire(NRF_SPIM_Type *bus);
+void nrf5x_spi_release(NRF_SPIM_Type *bus);
 
 /**
  * @brief USBDEV buffers must be word aligned because of DMA restrictions
diff --git a/cpu/nrf52/spi_twi_irq.c b/cpu/nrf52/spi_twi_irq.c
index 6db7b0065f..ca6868a88a 100644
--- a/cpu/nrf52/spi_twi_irq.c
+++ b/cpu/nrf52/spi_twi_irq.c
@@ -143,16 +143,22 @@ void spi_twi_irq_register_i2c(NRF_TWIM_Type *bus,
     NVIC_EnableIRQ(_isr[num]);
 }
 
-void nrf5x_i2c_acquire(NRF_TWIM_Type *bus)
+void nrf5x_i2c_acquire(NRF_TWIM_Type *bus,
+                       spi_twi_irq_cb_t cb, void *arg)
 {
     size_t num = _i2c_dev2num(bus);
     mutex_lock(&_locks[num]);
+    _irq[num] = cb;
+    _irq_arg[num] = arg;
 }
 
-void nrf5x_spi_acquire(NRF_SPIM_Type *bus)
+void nrf5x_spi_acquire(NRF_SPIM_Type *bus,
+                       spi_twi_irq_cb_t cb, void *arg)
 {
     size_t num = _spi_dev2num(bus);
     mutex_lock(&_locks[num]);
+    _irq[num] = cb;
+    _irq_arg[num] = arg;
 }
 
 void nrf5x_i2c_release(NRF_TWIM_Type *bus)
diff --git a/cpu/nrf5x_common/periph/i2c_nrf52_nrf9160.c b/cpu/nrf5x_common/periph/i2c_nrf52_nrf9160.c
index 2e3df5afb6..641d7aa70d 100644
--- a/cpu/nrf5x_common/periph/i2c_nrf52_nrf9160.c
+++ b/cpu/nrf5x_common/periph/i2c_nrf52_nrf9160.c
@@ -102,8 +102,13 @@ static void _init_pins(i2c_t dev)
 {
     gpio_init(i2c_config[dev].scl, GPIO_IN_OD_PU);
     gpio_init(i2c_config[dev].sda, GPIO_IN_OD_PU);
+}
+
+static void _setup_shared_peripheral(i2c_t dev)
+{
     bus(dev)->PSEL.SCL = i2c_config[dev].scl;
     bus(dev)->PSEL.SDA = i2c_config[dev].sda;
+    bus(dev)->FREQUENCY = i2c_config[dev].speed;
 }
 
 void i2c_init(i2c_t dev)
@@ -122,8 +127,8 @@ void i2c_init(i2c_t dev)
     /* configure pins */
     _init_pins(dev);
 
-    /* configure dev clock speed */
-    bus(dev)->FREQUENCY = i2c_config[dev].speed;
+    /* configure shared periphal speed */
+    _setup_shared_peripheral(dev);
 
     spi_twi_irq_register_i2c(bus(dev), i2c_isr_handler, (void *)(uintptr_t)dev);
 
@@ -161,8 +166,8 @@ void i2c_acquire(i2c_t dev)
     mutex_lock(&locks[dev]);
 
     if (IS_ACTIVE(MODULE_NRF5X_PERIPH_SHARED)) {
-        nrf5x_i2c_acquire(bus(dev));
-        _init_pins(dev);
+        nrf5x_i2c_acquire(bus(dev), i2c_isr_handler, (void *)(uintptr_t)dev);
+        _setup_shared_peripheral(dev);
     }
 
     bus(dev)->ENABLE = TWIM_ENABLE_ENABLE_Enabled;
diff --git a/cpu/nrf5x_common/periph/spi_nrf52_nrf9160.c b/cpu/nrf5x_common/periph/spi_nrf52_nrf9160.c
index fcd1e5ed69..fb1c792f5b 100644
--- a/cpu/nrf5x_common/periph/spi_nrf52_nrf9160.c
+++ b/cpu/nrf5x_common/periph/spi_nrf52_nrf9160.c
@@ -123,6 +123,13 @@ static void _clear_workaround(spi_t bus)
 #endif
 }
 
+static void _setup_shared_peripheral(spi_t bus)
+{
+    SPI_SCKSEL = spi_config[bus].sclk;
+    SPI_MOSISEL = spi_config[bus].mosi;
+    SPI_MISOSEL = spi_config[bus].miso;
+}
+
 void spi_init(spi_t bus)
 {
     assert(bus < SPI_NUMOF);
@@ -133,6 +140,7 @@ void spi_init(spi_t bus)
     mutex_lock(&busy[bus]);
     /* initialize pins */
     spi_init_pins(bus);
+    _setup_shared_peripheral(bus);
 }
 
 int spi_init_with_gpio_mode(spi_t bus, const spi_gpio_mode_t* mode)
@@ -165,9 +173,6 @@ void spi_init_pins(spi_t bus)
     spi_init_with_gpio_mode(bus, &gpio_modes);
 
     /* select pins for the SPI device */
-    SPI_SCKSEL = spi_config[bus].sclk;
-    SPI_MOSISEL = spi_config[bus].mosi;
-    SPI_MISOSEL = spi_config[bus].miso;
     _setup_workaround_for_ftpan_58(bus);
     spi_twi_irq_register_spi(dev(bus), spi_isr_handler, (void *)(uintptr_t)bus);
 }
@@ -180,8 +185,8 @@ void spi_acquire(spi_t bus, spi_cs_t cs, spi_mode_t mode, spi_clk_t clk)
     mutex_lock(&locks[bus]);
 
     if (IS_ACTIVE(MODULE_NRF5X_PERIPH_SHARED)) {
-        nrf5x_spi_acquire(dev(bus));
-        spi_init_pins(bus);
+        nrf5x_spi_acquire(dev(bus), spi_isr_handler, (void *)(uintptr_t)bus);
+        _setup_shared_peripheral(bus);
     }
 
     /* configure bus */
```

**Comment by maribu** [comment]
I was so free to add the comment to the `_setup_shared_peripheral()` function. Please squash. I think this is ready to go in.

**Comment by chrysn** [comment]
@nica-f confirmed earlier that it does now work, so her [comment to the contrary](https://github.com/RIOT-OS/RIOT/pull/18478#issuecomment-1223205961) should not stand in the way of this being merged.

**Review by maribu — APPROVED** [review]
Thx for addressing this. This will be highly useful not only for the Bangle.js 2, but also for many other nRF52 board (e.g. the BBC micro:bit v2, which already is sharing a peripheral for I2C and SPI, and consequently will have issues using both in the same app).

**Comment by maribu** [comment]
Would you mind to rephrase the commit title, as no module is introduced anymore?

Also, I don't think we need another full CI run to confirm that changing the commit message will not break anything.

**Comment by nica-f** [comment]
> @nica-f confirmed earlier that it does now work, so her [comment to the contrary](https://github.com/RIOT-OS/RIOT/pull/18478#issuecomment-1223205961) should not stand in the way of this being merged.

Yes, I can confirm this, sorry to be late to the party ;)
Works perfectly well, tested quite a bit on the Bangle.JS2 now where I2C sensors are spread on different pin sets, so far no issues, all sensors are accessible.

**Pull request merged**